GUI/PlotWidget.py

```diff
@@ -0,0 +1,281 @@
+				self.i = self.i-1
+				
+		#Add to plot and change intensity
+		self.scatter.set_offsets(np.c_[self.x_vals, self.y_vals])
+		self.intensity = np.concatenate((np.array(self.intensity) - (1 / self.colorsteps), new_tvals))
+		#print (self.intensity)
+		self.scatter.set_array(self.intensity)
+		
+		self.canvas.draw()
+		
+		return True
+				
+				
+	def update_occupancy_plot(self):
+		new_xvals, new_yvals, new_tvals = self.get_new_vals()
+		new_elements = np.c_[new_xvals, new_yvals]
+		self.occ_length.append(new_elements.shape[0])
+		self.old_elements = np.append(self.old_elements, new_elements, axis = 0)
+		
+		#count hited pixel
+		for new_element in new_elements:
+			pos = np.argwhere(np.all(self.occupancy_array[ : , :2] == new_element, axis = 1) == True)
+			if pos.size == 0:
+				# add new element
+				self.occupancy_array = np.append(self.occupancy_array, [np.append(new_element, 1)], axis = 0)
+				
+			elif pos.size == 1:
+				#increment element at pos
+				x = pos[0, 0]
+				self.occupancy_array[pos[0, 0], 2] = (self.occupancy_array[pos[0, 0], 2] + 1)
+				
+			else:
+				print("Error")
+				
+		#remove hited pixel
+		if self.j <= (self.integration_length):
+			self.j = self.j + 1
+			
+		elif self.j == (self.integration_length + 1):
+			number = self.occ_length[0]
+			self.occ_length.pop(0)
+			k = 0
+			while k < number:
+				k = k + 1
+				pos = np.argwhere(np.all(self.occupancy_array[ : , :2] == self.old_elements[1], axis = 1) == True)
+				self.old_elements = np.delete(self.old_elements, 1, axis = 0)
+				if self.occupancy_array[pos[0, 0], 2] == 1:
+					#Remove item if no count left
+					self.occupancy_array = np.delete(self.occupancy_array, pos[0, 0], axis = 0)
+					
+				elif self.occupancy_array[pos[0, 0], 2] > 1:
+					#decrement element at pos
+					self.occupancy_array[pos[0, 0], 2] = (self.occupancy_array[pos[0, 0], 2] - 1)
+					
+				else:
+					print("Error")
+					
+		elif self.j > (self.integration_length + 1):
+			while self.j > (self.integration_length + 1):
+				number = self.occ_length[0]
+				self.occ_length.pop(0)
+				k = 0
+				while k < number:
+					k = k + 1
+					pos = np.argwhere(np.all(self.occupancy_array[ : , :2] == self.old_elements[1], axis = 1) == True)
+					self.old_elements = np.delete(self.old_elements, 1, axis = 0)
+					if self.occupancy_array[pos[0, 0], 2] == 1:
+						#print("remove item if no count left")
+						self.occupancy_array = np.delete(self.occupancy_array,pos[0, 0], axis = 0)
+						
+					elif self.occupancy_array[pos[0, 0], 2] > 1:
+						#decrement element at pos
+						self.occupancy_array[pos[0, 0], 2] = (self.occupancy_array[pos[0, 0], 2] - 1)
+						
+					else:
+						print("Error")
+						
+				self.j = self.j - 1
+		
+		self.scatter.set_offsets(self.occupancy_array[ : , :2])
+		self.occupancy = self.occupancy_array[ : , 2: ]
+		self.scatter.set_array(np.squeeze(self.occupancy))
+		self.canvas.draw()
+		#print("update")
+		
+		return True
```

Comment:
With the above recommended changes to a full occupancy array, we can simplify this whole proc considerably.

```suggestion
    def update_occupancy_plot(self):
        new_xvals, new_yvals, new_tvals = self.get_new_vals()
        new_elements = np.c_[new_yvals, new_xvals]
        self.occ_length.append(new_elements.shape[0])
        self.old_elements = np.append(self.old_elements, new_elements, axis = 0)
        # use fancy indexing to add one to each hit pixel
        self.occupancy_array[new_yvals, new_xvals] += 1

       # first if branch is just the wait time until we covered one integration length
        if self.j <= (self.integration_length):
            self.j += 1
        else:
            # from here start rolling removal of old elements
            number = self.occ_length[0]
            self.occ_length.pop(0)
            toRemove = self.old_elements[0:number]
            self.occupancy_array[toRemove[:,0], toRemove[:,1]] -= 1
            self.old_elements = self.old_elements[number:, :]

        self.scatter.set_offsets(np.transpose(self.occupancy_array.nonzero()))
        self.scatter.set_array(self.occupancy_array[self.occupancy_array.nonzero()])
        self.canvas.draw()

        return True
```
This at the moment does not take into account a change to `integration_length`, but that's because on my machine it seems to reset the occupancy when I change the "plot depth" anyways. So I never run into the old last branch. Could be modified if that's actually required.